### DISCUSSION TITLE
How decryption with HashiCorp Vault work in kustomize-controller

### DISCUSSION THREAD

**Discussion by tz-torchai** [fluxcd/flux2#2721] [Q&A] [answered]
Hi,

In the [doc](https://fluxcd.io/docs/guides/mozilla-sops/#encrypting-secrets-using-hashicorp-vault), we can create a k8s secret from a vault token, which then will be used by ks-controller to send along with the decryption requests to vault. However, vault token usually has a [ttl](https://learn.hashicorp.com/tutorials/vault/tokens), especially in this case it should have a ttl. Therefore, after the expiration time, kustomize-controller won't be able to decrypt the secret. 

Is my above statement correct?

If it's correct, then we should figure out some way to dynamically retrieve the token, e.g. doing kubernetes auth using service account.

**Comment by souleb** [comment, marked as answer]
We use Mozilla sops for this feature.

Sops uses enveloppe encryption, i.e. it uses a hc vault key as master key (multiple keys are supported) to encrypt a data key which itself is used to encrypt the sensitive data.

I have not tested it but sops has a way to [update keys]( https://github.com/mozilla/sops#id12) and [rotate data keys](https://github.com/mozilla/sops#id19).

This [doc](https://www.hashicorp.com/blog/manage-kubernetes-secrets-for-flux-with-hashicorp-vault) provides alternatives.

**Reply by tz-torchai** [reply]
Thanks! This really help clarify things! 

But based on the new info, ks-controller still needs to send decryption requests to vault to get decrypted data key (since master key is stored in vault). Then ks-controller will use this data key to decrypt the ciphertext locally.

Therefore ks-controller still needs a valid vault token in order to talk to vault.

**Reply by tz-torchai** [reply]
@souleb  After reading ks-controller code and test, i find it still needs [vault token](https://github.com/fluxcd/kustomize-controller/blob/9db141d9dbeb533295b378112558362f5d8dbbeb/internal/sops/keyservice/server_test.go#L105) during the decryption of data key. 

Therefore I think ks-controller still needs to authenticate against vault so it can decrypt the data key.

**Reply by tz-torchai** [reply]
Another thing I am curiours about is, where is the function call / logic to actually decrypt the cipher text (**the actual string data**) in the k8s secret using data key? I could only find the logic to decrypt **data key** in ks-controller.

**Reply by hiddeco** [reply]
The logic to decrypt the actual data is in https://github.com/fluxcd/kustomize-controller/blob/main/controllers/kustomization_decryptor.go. This makes use of a custom local SOPS keyservice server implementation for every Kustomization object constructed with the credentials specific to this object (https://github.com/fluxcd/kustomize-controller/blob/main/internal/sops/keyservice/server.go).

There is at present an expectation for something else to rotate this token if it has a TTL.

**Reply by tz-torchai** [reply]
Thanks, I read through all related files and logic regarding the decryption (both in sops and ks-controller). If I understand it correctly, [this](https://github.com/fluxcd/kustomize-controller/blob/main/controllers/kustomization_decryptor.go#L258) should be the line to decrypt the actual data with data key, it [does so](https://github.com/mozilla/sops/blob/v3.7.3/sops.go#L374) by directly modifying the `sops.Tree` object.


Back to the original issue, IMO, the token used for encryption shouldn't be used for decryption, since the token was obtained after authentication by a human user, which should always have a short TTL. But even if it doesn't expire, a vault token shouldn't be present to two different users(a human user and a machine user).


I think we can use https://www.vaultproject.io/docs/auth/kubernetes or JWT Auth to let ks-controller obtain the token, and don't even create the `sops-hcvault` secret since it will never be used.
